src/translation.cpp

```diff
@@ -143,14 +220,338 @@ void Translations::loadTranslation(const std::string &data)
-		if (!oword2.empty()) {
-			std::wstring translation_index = textdomain + L"|";
-			translation_index.append(oword1);
-			m_translations[translation_index] = oword2;
-		} else {
-			infostream << "Ignoring empty translation for \""
-				<< wide_to_utf8(oword1) << "\"" << std::endl;
+		addTranslation(textdomain, word1.str(), word2.str());
+	}
+}
+
+
+std::wstring Translations::unescapeC(const std::wstring &str) {
+	// Process escape sequences in str as if it were a C string
+	std::wstring result;
+	size_t i = 0;
+	while (i < str.length()) {
+		if (str[i] != L'\\') {
+			result.push_back(str[i]);
+			i++;
+			continue;
+		}
+		i++;
+		if (i == str.length()) {
+			errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+			break;
+		}
+		switch (str[i]) {
+			// From https://en.wikipedia.org/wiki/Escape_sequences_in_C#Table_of_escape_sequences
+			case L'a': result.push_back(L'\a'); break;
+			case L'b': result.push_back(L'\b'); break;
+			case L'e': result.push_back(L'\x1b'); break;
+			case L'f': result.push_back(L'\f'); break;
+			case L'n': result.push_back(L'\n'); break;
+			case L'r': result.push_back(L'\r'); break;
+			case L't': result.push_back(L'\t'); break;
+			case L'v': result.push_back(L'\v'); break;
+			case L'\\': result.push_back(L'\\'); break;
+			case L'\'': result.push_back(L'\''); break;
+			case L'"': result.push_back(L'"'); break;
+			case L'?': result.push_back(L'?'); break;
+			case L'0': case L'1': case L'2': case L'3': case L'4': case L'5': case L'6': case L'7': {
+				size_t j = 0;
+				wchar_t c = 0;
+				for (; j < 3 && i+j < str.length() && L'0' <= str[i+j] && str[i+j] <= L'7'; j++) {
+					c = c * 8 + (str[i+j] - L'0');
+				}
+				if (c <= 0xff) {
+					result.push_back(c);
+				}
+				i += j;
+				continue;
+			}
+			case L'x': {
+				i++;
+				if (i >= str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char32_t c = 0;
+				size_t j = 0;
+				unsigned char v;
+				for (; i+j < str.length() && hex_digit_decode((char)str[i+j], v); j++) {
+					c = c * 16 + v;
+				}
+				if (j == 0) {
+					errorstream << "Invalid escape sequence \\x, ignoring" << std::endl;
+					continue;
+				}
+				// If character fits in 16 bits and is not part of surrogate pair, insert it.
+				// Otherwise, silently drop it: this is valid since \x escape sequences with
+				// values above 0xff are implementation-defined
+				if ((c < 0xd800) || (0xe000 <= c && c <= 0xffff)) {
+					result.push_back(c);
+				}
+				i += j;
+				continue;
+			}
+			case L'u': {
+				i++;
+				if (i + 4 > str.length()) {
+					errorstream << "Unfinished escape sequence at the end of \"" << wide_to_utf8(str) << "\"" << std::endl;
+				}
+				char16_t c = 0;
+				bool ok = true;
+				for (size_t j = 0; j < 4; j++) {
+					unsigned char v;
+					if (str[i+j] <= 0xff && hex_digit_decode((char)str[i+j], v)) {
+						c = c * 16 + v;
```

Comment:
I know it's pretty simple but do we really need to reimplement number parsing here?

If it's kept it should at least be refactored into a helper function (I suggest an inline lambda since it's only needed by this function).